### PR TITLE
Support mnnvl all2allv from Flashinfer

### PR DESCRIPTION
**Pull request by wenscarl**
## Essential Elements of an Effective PR Description Checklist
- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

Needs https://github.com/flashinfer-ai/flashinfer/pull/1245

## Test Plan

```
VLLM_ALL2ALL_BACKEND="flashinfer_all2allv" \
CUDA_VISIBLE_DEVICES=0,1,2,3 \
VLLM_USE_FLASHINFER_MOE_FP4=1 \
VLLM_FLASHINFER_MOE_BACKEND="throughput" \
  /home/shuw/.local/bin/vllm serve nvidia/DeepSeek-R1-FP4 \
    --quantization="modelopt_fp4" \
    --trust-remote-code \
    --max-model-len=2048 \
    --block-size=128 \
    --max-num-seqs=256 \
    --enable-expert-parallel \
    --gpu_memory_utilization=0.8 \
    --tensor-parallel-size 1 \
    --data-parallel-size 4
    
python benchmarks/benchmark_serving.py \
  --model nvidia/DeepSeek-R1-FP4 \
  --dataset-name random \
  --ignore-eos \
  --num-prompts 256 \
  --max-concurrency 256 \
  --random-input-len 128 \
  --random-output-len 1024 
```

vs. 
VLLM_ALL2ALL_BACKEND="naive" \

... 
## Test Result
accuracy:
```
VLLM_USE_FLASHINFER_MOE_FP4=1 VLLM_FLASHINFER_MOE_BACKEND="throughput" VLLM_WORKER_MULTIPROC_METHOD=spawn VLLM_ALL2ALL_BACKEND="flashinfer_all2allv" \
lm_eval --model vllm --model_args pretrained=nvidia/DeepSeek-R1-FP4,quantization=modelopt_fp4,data_parallel_size=4,enable_expert_parallel=True,tensor_parallel_size=1,enforce_eager=True,max_model_len=2048 --trust_remote_code --tasks gsm8k --num_fewshot 5 --batch_size auto

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9371|±  |0.0067|
|     |       |strict-match    |     5|exact_match|↑  |0.9340|±  |0.0068|
```

perf:
Alltoallv(this PR):
```
Successful requests:                     256       
Benchmark duration (s):                  111.43    
Total input tokens:                      32512     
Total generated tokens:                  262144    
Request throughput (req/s):              2.30      
Output token throughput (tok/s):         2352.63   
Total Token throughput (tok/s):          2644.42   
---------------Time to First Token----------------
Mean TTFT (ms):                          9734.68   
Median TTFT (ms):                        9769.39   
P99 TTFT (ms):                           9794.07   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          99.37     
Median TPOT (ms):                        99.34     
P99 TPOT (ms):                           99.35     
---------------Inter-token Latency----------------
Mean ITL (ms):                           99.37     
Median ITL (ms):                         90.51     
P99 ITL (ms):                            143.30    
```

allgather-reducescatter
```
Successful requests:                     256       
Benchmark duration (s):                  134.83    
Total input tokens:                      32512     
Total generated tokens:                  262144    
Request throughput (req/s):              1.90      
Output token throughput (tok/s):         1944.28   
Total Token throughput (tok/s):          2185.42   
---------------Time to First Token----------------
Mean TTFT (ms):                          9614.91   
Median TTFT (ms):                        9648.26   
P99 TTFT (ms):                           9673.86   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          122.36    
Median TPOT (ms):                        122.33    
P99 TPOT (ms):                           122.34    
---------------Inter-token Latency----------------
Mean ITL (ms):                           122.36    
Median ITL (ms):                         122.26    
P99 ITL (ms):                            129.50 
```

### CODE REVIEW

File: vllm/envs.py [hunk 1]
```diff
@@ -1040,6 +1040,7 @@ def get_vllm_port() -> Optional[int]:
     # - "pplx": use pplx kernels
     # - "deepep_high_throughput", use deepep high-throughput kernels
     # - "deepep_low_latency", use deepep low-latency kernels
+    # - "flashinfer", use flashinfer alltoallv kernels for mnnvl
```
Comment (tlrmchlsmth) [inline vllm/envs.py]: Should we be more explicit here in case flashinfer adds other all2all kernels that have a different interface in the future?

```suggestion
    # - "flashinfer_alltoallv", use flashinfer alltoallv kernels for mnnvl
```

File: vllm/distributed/device_communicators/all2all.py [hunk 2]
```diff
@@ -47,11 +54,20 @@ def dispatch(self, hidden_states: torch.Tensor,
                  router_logits: torch.Tensor):
         cu_tokens_across_dp_cpu = get_forward_context(
         ).dp_metadata.cu_tokens_across_dp_cpu
+        if not use_ag_rs:
+            hidden_states = self.naive_multicast(hidden_states,
+                                                cu_tokens_across_dp_cpu)
+            router_logits = self.naive_multicast(router_logits,
+                                                cu_tokens_across_dp_cpu)
```
Comment (tlrmchlsmth) [inline vllm/distributed/device_communicators/all2all.py]: Should we go ahead and delete the `naive_multicast` here?

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 3]
```diff
@@ -3,20 +3,36 @@
...
 from vllm.utils.flashinfer import nvfp4_block_scale_interleave
 
 
+def get_global_num_tokens_cpu():
+    cu_sizes = get_forward_context().dp_metadata.cu_tokens_across_dp_cpu
+    sizes = [cu_sizes[0].item()]
+    for i in range(1, len(cu_sizes)):
+        sizes.append((cu_sizes[i] - cu_sizes[i - 1]).item())
+    return sizes
+
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Looks like this should be removed. It returns the same thing as `get_local_sizes`, and isn't used anywhere

File: vllm/model_executor/layers/quantization/modelopt.py
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/modelopt.py:1]: Could you revert the changes to this file?

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 4]
```diff
@@ -92,6 +141,103 @@ def finalize(self, output: torch.Tensor, fused_expert_output: torch.Tensor,
...
-                fused_expert_output, dim=0, sizes=get_local_sizes())
-        output.copy_(fused_expert_output)
+            if enable_flashinfer_fp4_allgather:
+                fused_expert_output = get_dp_group().reduce_scatterv(
+                    fused_expert_output, dim=0, sizes=get_local_sizes())
+                output.copy_(fused_expert_output)
+
+            if enable_flashinfer_alltoall:
+                all2all_manager = get_ep_group(
+                ).device_communicator.all2all_manager
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Can this be looked up in `__init__` and stashed in a member var?

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 5]
```diff
@@ -92,6 +141,103 @@ def finalize(self, output: torch.Tensor, fused_expert_output: torch.Tensor,
...
+                fused_expert_output = flashinfer_alltoall_combine(
+                    all2all_manager,
+                    fused_expert_output,
+                    top_k=top_k,
+                    token_count=token_count,
+                    alltoall_info=self.alltoall_info,
+                )
+                output.copy_(fused_expert_output)
+        else:
+            output.copy_(fused_expert_output)
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Why not let all the above cases fall through to the end and have a single copy?

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 6]
```diff
@@ -3,20 +3,36 @@
...
+        sizes.append((cu_sizes[i] - cu_sizes[i - 1]).item())
+    return sizes
+
+
 def get_local_sizes():
     return get_forward_context().dp_metadata.get_chunk_sizes_across_dp_rank()
 
 
+enable_flashinfer_alltoall = envs.VLLM_ALL2ALL_BACKEND == "flashinfer"
+enable_flashinfer_fp4_allgather = not enable_flashinfer_alltoall
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Why not have two separate `PrepareaAndFinalize` classes if these conditions are mutually exclusive?

**Comment by mgoin** [comment]
@wenscarl Are you waiting for additional review on this?

**Comment by wenscarl** [comment]
> @wenscarl Are you waiting for additional review on this?

PTAL. Thanks!

File: vllm/distributed/device_communicators/all2all.py
Comment (bnellnm) [inline vllm/distributed/device_communicators/all2all.py:1]: Is the `naive_multicast` method needed anymore?  Seems like there are no uses now.

File: vllm/distributed/device_communicators/all2all.py [hunk 7]
```diff
@@ -291,3 +296,92 @@ def get_handle(self, kwargs):
...
+        self.alltoall_info = None
+
+    def initialize(
+            self,
+            world_size: int,
+            rank: int,
+            gpus_per_node: int,
+    ):
+        """Initialize workspace"""
+        if self.initialized and self.world_size == world_size:
```
Comment (bnellnm) [inline vllm/distributed/device_communicators/all2all.py]: When would the world size change?

Comment (wenscarl) [reply vllm/distributed/device_communicators/all2all.py]: Hmm, It's not necessary. Will delete.

File: vllm/distributed/device_communicators/all2all.py [hunk 8]
```diff
@@ -291,3 +277,92 @@ def get_handle(self, kwargs):
...
+
+    def get_handle(self, kwargs):
+        return self
+
+    def cleanup(self):
+        """Clean up workspace"""
+        if self.initialized and self.workspace_tensor is not None \
+            and self.prepare_workspace_tensor is not None:
+            try:
+                del self.workspace_tensor
```
Comment (bnellnm) [inline vllm/distributed/device_communicators/all2all.py:390]: does the `prepare_workspace_tensor` need to be deleted also?

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 9]
```diff
@@ -1433,6 +1433,7 @@ def apply(
             routing_method_type = flashinfer.RoutingMethodType.DeepSeekV3
             if use_llama4_routing:
                 routing_method_type = flashinfer.RoutingMethodType.Llama4
+            # out = torch.empty_like(x)
```
Comment (bnellnm) [inline vllm/model_executor/layers/quantization/modelopt.py]: remove debug cruft.

**Comment by bnellnm** [comment]
LGTM.  Just had a few minor comments.

**Review by bnellnm — COMMENTED** [review]
I think this is good to go once the lint/test issues are fixed.  Not sure what's going on with the cuda-python package.

**Review by bnellnm — APPROVED** [review]

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 10]
```diff
@@ -3,9 +3,12 @@
 from typing import Optional
 
 import torch
+from flashinfer.comm.trtllm_alltoall import MnnvlMoe, MoEAlltoallInfo
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Maybe this import needs to be moved into `flashinfer_alltoall_dispatch` to avoid the test failures in CI?

**Review by bnellnm — APPROVED** [review]

File: vllm/distributed/device_communicators/all2all.py [hunk 11]
```diff
@@ -26,44 +32,24 @@ class NaiveAll2AllManager(All2AllManagerBase):
...
-        start = 0 if self.dp_rank == 0 else cu_tokens_across_dp_cpu[
-            self.dp_rank - 1]
-        end = cu_tokens_across_dp_cpu[self.dp_rank]
-        buffer[start:end, :].copy_(x)
-        for idx in range(self.dp_world_size):
-            start = 0 if idx == 0 else cu_tokens_across_dp_cpu[idx - 1]
-            end = cu_tokens_across_dp_cpu[idx]
-            self.dp_group.broadcast(buffer[start:end, :], idx)
-
-        return buffer
```
Comment (tlrmchlsmth) [inline vllm/distributed/device_communicators/all2all.py:46]: Let's keep the naive_multicast for this PR. Now that we default to `allgather_reducescatter`, I don't think there's a big value in doing this. And I'd like to have at least a release where we deprecate the naive implementation before we remove it, in case the `all_gatherv`/`reduce_scatterv` implementation doesn't work in some cases

File: vllm/distributed/device_communicators/mnnvl_compat.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+# from vllm.distributed import get_dp_group
```
Comment (tlrmchlsmth) [inline vllm/distributed/device_communicators/mnnvl_compat.py]: nit: remove commented out import

**Review by tlrmchlsmth — COMMENTED** [review]
Please revert the changes to the naive all2all backend for now and then LGTM!

File: vllm/distributed/device_communicators/mnnvl_compat.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        dist.all_gather_object(gathered, data, group=self._group)
+        return gathered
+
+    def Split(self, color: int, key: int) -> 'CustomCommunicator':
+        return self
+
+    def allgather_bytes(self, data: bytes):
+        result = [data] * self.Get_size()
+        torch.distributed.all_gather_object(result, data)
+        return result
```
Comment (tlrmchlsmth) [inline vllm/distributed/device_communicators/mnnvl_compat.py]: Making one last pass and this caught my eye. Can we remove `Split` and `allgather_bytes`? I don't see them being used anywhere.

Seems like may be are used by `flashinfer.comm.mnnvl.CommBackend`, but why does the capitalization differ?

Comment (tlrmchlsmth) [reply vllm/distributed/device_communicators/mnnvl_compat.py]: I don't see any `allgather_bytes` in https://github.com/flashinfer-ai/flashinfer, so I'm going to remove it

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by mgoin** [comment]
@wenscarl can you fix the latest merge conflict?
